### PR TITLE
implement receive window auto tuning

### LINKED ISSUE #9: Dynamic Window Sizing
Currently, yamux uses a fixed-size window. Unfortunately, this means we can't account for variable latencies. For low-latency connections, the window is too large and we'll have to hold large buffers. For high-latency connections, the window is too small and we'll bottleneck on filling up the window.

Ideally, we'd extract the QUIC dynamic window sizing logic from https://github.com/lucas-clemente/quic-go and use it to figure out our window sizing.

### PR DESCRIPTION
**Pull request by marten-seemann**
Fixes #9.

Based on the algorithm used in QUIC: https://docs.google.com/document/d/1F2YfdDXKpy20WVKJueEf4abn_LVZHhMUMS5gX6Pgjl4/edit#heading=h.hcm2y5x4qmqt.

What we do is compare the frequency we're updating the receive window. Updates happen when we've consumed more than half the available. If we're updating more often than once every 2 RTTs, we double the window size (up to a configurable maximum window size).
Unfortunately, we don't have a good way to get an up-to-date RTT measurement here. Instead, we use the ping function to generate an estimate right when we initialize yamux, and use that estimate for the rest of the connection.

### CODE REVIEW

**Comment by willscott** [comment]
```
session_test.go:1237: sendWindow: exp=32768, got=98304
--- FAIL: TestSession_PartialReadWindowUpdate (1.02s)
```

(I didn't look to see if that flakiness was already present.)

**Comment by marten-seemann** [comment]
@willscott Not really happy with my fix (or the test case as a whole, for that matter), but at least it shouldn't be flaky any more.

**Review by willscott — APPROVED** [review]

**Review by Stebalien — CHANGES_REQUESTED** [review]

File: util.go [hunk 1]
```diff
@@ -88,20 +102,37 @@ func (s *segmentedBuffer) Len() uint32 {
 
 // If the space to write into + current buffer size has grown to half of the window size,
 // grow up to that max size, and indicate how much additional space was reserved.
-func (s *segmentedBuffer) GrowTo(max uint32, force bool) (bool, uint32) {
+func (s *segmentedBuffer) GrowTo(force bool, now time.Time) (bool, uint32) {
```
Comment (Stebalien) [inline util.go:105]: This logic doesn't really belong in this type. Can't we compute the target window outside this type and pass it in?

**Comment by aschmahmann** [comment]
@marten-seemann what's the status of this PR?

**Comment by marten-seemann** [comment]
@aschmahmann It's in the very unfortunate status of "submitted PR, received review, and totally forgot about it". Sorry for that! I'll clean it up today.

**Pull request closed without merging**